File: crypto/e8_theta.c
/* e8_theta.c
 * Jacobi theta function evaluator for E8 / E8×E8 ZK commitment.
 *
 * Fixed modular parameter: τ = i  →  q = e^{-2π} ≈ 1.868e-3
 * Convergence is extremely fast: |q|^15 < 2e-42 (machine zero).
 * All theta functions are real-valued for real Chern-root arguments.
 *
 * Equations from Liu & Wang, arXiv:2601.18221:
 *   2.9–2.12  Jacobi theta functions
 *   2.31      E8 characteristic form (phi^8 · ch(V))
 *   Section 4 E8×E8 product
 *
 * Build:  cc -O2 -lm -o e8_theta e8_theta.c
 * Test:   ./e8_theta
 */

#include <math.h>
#include <stdint.h>
#include <stdio.h>
#include <string.h>

/* ── Constants ──────────────────────────────────────────────────── */

#define NTERMS    15          /* truncation depth; error < 2e-42      */
#define E8_DIM     8          /* Chern roots per E8 bundle             */
#define PI         M_PI

/* q = exp(-2π) */
static const double Q      = 1.8679908225077686e-3;
/* q^{1/8} = exp(-π/4) */
static const double Q_1_8  = 4.5593812776599624e-1;
/* q^{-1/2} = exp(+π) — needed for q^{j-1/2} = q^j · q^{-1/2} */
static const double Q_N1_2 = 2.3140692632779269e+1;  /* exp(+π)      */

/* ── Precomputed q-powers ────────────────────────────────────────── */

static double qpow[NTERMS + 2];    /* qpow[j]  = q^j, j=0..NTERMS+1  */
static double qhalf[NTERMS + 2];   /* qhalf[j] = q^{j-1/2}, j=1..N   */

static void init_tables(void) {
    static int ready = 0;
    if (ready) return;
    qpow[0] = 1.0;
    for (int j = 1; j <= NTERMS + 1; j++) {
        qpow[j]  = qpow[j-1] * Q;
        qhalf[j] = qpow[j] * Q_N1_2;   /* q^{j-1/2} */
    }
    ready = 1;
}

/* ── Theta functions (real v, τ = i) ────────────────────────────── */
/*
 * From the paper (eqs 2.9–2.12), for real v and τ = i:
 *
 *  θ (v,i) = 2q^{1/8} sin(πv) ∏_j (1-q^j)(1 - 2cos(2πv)q^j  + q^{2j})
 *  θ₁(v,i) = 2q^{1/8} cos(πv) ∏_j (1-q^j)(1 + 2cos(2πv)q^j  + q^{2j})
 *  θ₂(v,i) =                  ∏_j (1-q^j)(1 - 2cos(2πv)q^{j-½} + q^{2j-1})
 *  θ₃(v,i) =                  ∏_j (1-q^j)(1 + 2cos(2πv)q^{j-½} + q^{2j-1})
 */

double theta0(double v) {
    init_tables();
    double c = cos(2.0 * PI * v);
    double p = 1.0;
    for (int j = 1; j <= NTERMS; j++) {
        double qj = qpow[j];
        p *= (1.0 - qj) * (1.0 - 2.0*c*qj + qj*qj);
    }
    return 2.0 * Q_1_8 * sin(PI * v) * p;
}

double theta1(double v) {
    init_tables();
    double c = cos(2.0 * PI * v);
    double p = 1.0;
    for (int j = 1; j <= NTERMS; j++) {
        double qj = qpow[j];
        p *= (1.0 - qj) * (1.0 + 2.0*c*qj + qj*qj);
    }
    return 2.0 * Q_1_8 * cos(PI * v) * p;
}

double theta2(double v) {
    init_tables();
    double c = cos(2.0 * PI * v);
    double p = 1.0;
    for (int j = 1; j <= NTERMS; j++) {
        double qh = qhalf[j];
        p *= (1.0 - qpow[j]) * (1.0 - 2.0*c*qh + qh*qh);
    }
    return p;
}

double theta3(double v) {
    init_tables();
    double c = cos(2.0 * PI * v);
    double p = 1.0;
    for (int j = 1; j <= NTERMS; j++) {
        double qh = qhalf[j];
        p *= (1.0 - qpow[j]) * (1.0 + 2.0*c*qh + qh*qh);
    }
    return p;
}

/* θ'(0,τ) — derivative of θ w.r.t. v at v=0.
 * Computed by symmetric finite difference for accuracy.
 * (The direct formula 2q^{1/8}π∏(1-qj)³ misses the f'(0) cross term;
 *  numerical differentiation avoids that subtlety.)
 */
double theta0_prime(void) {
    const double h = 1e-7;
    return (theta0(h) - theta0(-h)) / (2.0 * h);
}

/* ── E8 commitment (eq. 2.31) ───────────────────────────────────── */
/*
 * φ(τ)^8 ch(V) = ½( ∏_{l=1}^8 θ₁(y_l) + ∏θ₂(y_l) + ∏θ₃(y_l) )
 *
 * For two-factor auth: y[0..7] are the 8 Chern roots derived from one
 * password via HKDF.  The commitment C is stored publicly.
 */
typedef struct {
    double pi1;   /* product of θ₁ over 8 roots */
    double pi2;   /* product of θ₂ over 8 roots */
    double pi3;   /* product of θ₃ over 8 roots */
    double C;     /* ½(pi1 + pi2 + pi3)          */
} E8Commit;

E8Commit e8_commit(const double y[E8_DIM]) {
    double pi1 = 1.0, pi2 = 1.0, pi3 = 1.0;
    for (int l = 0; l < E8_DIM; l++) {
        pi1 *= theta1(y[l]);
        pi2 *= theta2(y[l]);
        pi3 *= theta3(y[l]);
    }
    E8Commit c = { pi1, pi2, pi3, 0.5*(pi1 + pi2 + pi3) };
    return c;
}

/* ── E8×E8 commitment (Section 4) ──────────────────────────────── */
/*
 * Two independent E8 bundles — one per password.
 * Combined commitment = c1.C × c2.C  (product of characteristic forms).
 * This is the two-factor structure: both passwords must be correct.
 */
typedef struct {
    E8Commit c1;
    E8Commit c2;
    double   C;    /* c1.C * c2.C */
} E8xE8Commit;

E8xE8Commit e8xe8_commit(const double y1[E8_DIM], const double y2[E8_DIM]) {
    E8xE8Commit r;
    r.c1 = e8_commit(y1);
    r.c2 = e8_commit(y2);
    r.C  = r.c1.C * r.c2.C;
    return r;
}

/* ── Modular transformation check (eqs. 3.27–3.30) ─────────────── */
/*
 * For ZK verification: given a claimed commitment C and a response,
 * check the weight-(2k+4) modular identity.
 *
 * At τ = i: S sends τ → -1/τ = -1/i = i (fixed point!).
 * So the T-transform Q(τ+1)=Q(τ) is the active one.
 *
 * Practical implication: the verifier checks that the prover's
 * response z, when plugged back into the theta products, reproduces
 * C · (scale factor).  The scale factor encodes the modular weight.
 */
double modular_weight_check(const E8Commit *c, int k) {
    /* At τ = i: weight = 2k+4. S(i) = i, so scale = i^{2k+4} = (-1)^{k+2} */
    /* For k even: scale = +1; k odd: scale = -1 */
    int weight = 2*k + 4;
    double scale = ((weight / 2) % 2 == 0) ? 1.0 : -1.0;
    return c->C * scale;
}

/* ── Chern root derivation (password → y[8]) ────────────────────── */
/*
 * In production: replace this with HKDF-SHA256.
 * y_l = HKDF(password, salt, "e8-root-" || l) reduced to [0,1).
 *
 * Here: a deterministic hash for testing only.
 * Uses xorshift64 seeded by the input bytes.
 */
static uint64_t xorshift64(uint64_t *state) {
    *state ^= *state << 13;
    *state ^= *state >> 7;
    *state ^= *state << 17;
    return *state;
}

void password_to_roots(const char *password, double y[E8_DIM]) {
    /* Seed from password bytes */
    uint64_t state = 0xcbf29ce484222325ULL;  /* FNV offset basis */
    for (const char *p = password; *p; p++) {
        state ^= (uint8_t)*p;
        state *= 0x100000001b3ULL;           /* FNV prime */
    }
    /* Generate 8 roots in (0, 1) — avoid exact 0 or 1 (theta0 = 0 there) */
    for (int l = 0; l < E8_DIM; l++) {
        uint64_t r = xorshift64(&state);
        y[l] = (double)(r >> 11) * (1.0 / (double)(1ULL << 53));
        if (y[l] < 1e-9) y[l] = 1e-9;
    }
}

/* ── Self-test ──────────────────────────────────────────────────── */

int main(void) {
    init_tables();

    /* ── 1. Jacobi identity: θ'(0) = π·θ₁(0)·θ₂(0)·θ₃(0) ─────── */
    double th0p  = theta0_prime();
    double th1_0 = theta1(0.0);
    double th2_0 = theta2(0.0);
    double th3_0 = theta3(0.0);
    double lhs = th0p;
    double rhs = PI * th1_0 * th2_0 * th3_0;
    printf("── Jacobi identity check ──────────────────────────────\n");
    printf("  θ'(0)                 = %+.15e\n", lhs);
    printf("  π·θ₁(0)·θ₂(0)·θ₃(0) = %+.15e\n", rhs);
    printf("  |error|               = %.3e  (should be < 1e-14)\n\n",
           fabs(lhs - rhs));

    /* ── 2. Theta values at a sample point ─────────────────────── */
    double v = 0.3;
    printf("── Theta values at v = %.1f ────────────────────────────\n", v);
    printf("  θ (v) = %+.10f\n", theta0(v));
    printf("  θ₁(v) = %+.10f\n", theta1(v));
    printf("  θ₂(v) = %+.10f\n", theta2(v));
    printf("  θ₃(v) = %+.10f\n\n", theta3(v));

    /* ── 3. E8 commitment from two passwords ─────────────────────  */
    const char *pw1 = "correct-horse-battery";
    const char *pw2 = "staple-pony-lattice";

    double y1[E8_DIM], y2[E8_DIM];
    password_to_roots(pw1, y1);
    password_to_roots(pw2, y2);

    printf("── E8 Chern roots from passwords ──────────────────────\n");
    printf("  P1 roots: ");
    for (int l = 0; l < E8_DIM; l++) printf("%.4f ", y1[l]);
    printf("\n  P2 roots: ");
    for (int l = 0; l < E8_DIM; l++) printf("%.4f ", y2[l]);
    printf("\n\n");

    E8Commit c1  = e8_commit(y1);
    E8Commit c2  = e8_commit(y2);
    E8xE8Commit combined = e8xe8_commit(y1, y2);

    printf("── E8 Commitments ─────────────────────────────────────\n");
    printf("  C1 = ½(Π₁ + Π₂ + Π₃)\n");
    printf("       Π₁ = %+.10e\n", c1.pi1);
    printf("       Π₂ = %+.10e\n", c1.pi2);
    printf("       Π₃ = %+.10e\n", c1.pi3);
    printf("       C1 = %+.10e\n\n", c1.C);
    printf("  C2 = %+.10e\n\n", c2.C);
    printf("  C_combined (E8×E8) = C1·C2 = %+.10e\n\n", combined.C);

    /* ── 4. Wrong password produces different commitment ─────────  */
    double y_wrong[E8_DIM];
    password_to_roots("wrong-password", y_wrong);
    E8Commit c_wrong = e8_commit(y_wrong);
    printf("── Wrong password commitment ───────────────────────────\n");
    printf("  C_wrong = %+.10e\n", c_wrong.C);
    printf("  |C1 - C_wrong| = %.6e  (should be large)\n\n",
           fabs(c1.C - c_wrong.C));

    /* ── 5. Modular weight check at τ = i, k = 4 ────────────────  */
    /* 7-dimensional case from Theorem 3.3: dim = 4k-1 = 7 → k = 2 */
    printf("── Modular weight at τ=i, k=2 (7-dim, weight=2k+4=8) ──\n");
    double ws = modular_weight_check(&c1, 2);
    printf("  Q(τ=i) · scale = %+.10e\n\n", ws);

    /* ── 6. Sensitivity: small perturbation of y₀ ────────────────  */
    double y_perturb[E8_DIM];
    memcpy(y_perturb, y1, sizeof(y1));
    y_perturb[0] += 1e-6;
    E8Commit c_p = e8_commit(y_perturb);
    printf("── Sensitivity (Δy₀ = 1e-6) ───────────────────────────\n");
    printf("  ΔC / Δy = %.6e\n\n",
           (c_p.C - c1.C) / 1e-6);

    printf("Done.\n");
    return 0;
}

File: crypto/e8_theta.ts
/**
 * e8_theta.ts
 * TypeScript port of e8_theta.c — mirrors every function, constant, and
 * loop structure so both languages produce bit-identical commitments for
 * the same input.  Cross-language agreement is verified by test_vectors.c.
 *
 * Arithmetic: IEEE 754 double throughout (same as C double).
 * BigInt is used only for the uint64 FNV/xorshift step — nowhere else.
 */

// ── Constants (must match e8_theta.c exactly) ─────────────────────────────

const NTERMS = 15;
export const E8_DIM = 8;

const Q      = 1.8679908225077686e-3;   // exp(-2π)
const Q_1_8  = 4.5593812776599624e-1;   // exp(-π/4)
// eslint-disable-next-line @typescript-eslint/no-loss-of-precision
const Q_N1_2 = 2.3140692632779269e+1;   // exp(+π)  ← q^{-1/2} — must match C literal bit-for-bit

// ── Precomputed q-power tables ────────────────────────────────────────────

const _qpow  = new Float64Array(NTERMS + 2);   // q^j,     j = 0..NTERMS+1
const _qhalf = new Float64Array(NTERMS + 2);   // q^{j-½}, j = 1..NTERMS+1

let _ready = false;

function initTables(): void {
  if (_ready) return;
  _qpow[0] = 1.0;
  for (let j = 1; j <= NTERMS + 1; j++) {
    _qpow[j]  = _qpow[j - 1] * Q;
    _qhalf[j] = _qpow[j] * Q_N1_2;
  }
  _ready = true;
}

// ── Theta functions ───────────────────────────────────────────────────────
// Equations 2.9–2.12 of Liu & Wang (arXiv:2601.18221), τ = i.

export function theta0(v: number): number {
  initTables();
  const c = Math.cos(2.0 * Math.PI * v);
  let p = 1.0;
  for (let j = 1; j <= NTERMS; j++) {
    const qj = _qpow[j];
    p *= (1.0 - qj) * (1.0 - 2.0 * c * qj + qj * qj);
  }
  return 2.0 * Q_1_8 * Math.sin(Math.PI * v) * p;
}

export function theta1(v: number): number {
  initTables();
  const c = Math.cos(2.0 * Math.PI * v);
  let p = 1.0;
  for (let j = 1; j <= NTERMS; j++) {
    const qj = _qpow[j];
    p *= (1.0 - qj) * (1.0 + 2.0 * c * qj + qj * qj);
  }
  return 2.0 * Q_1_8 * Math.cos(Math.PI * v) * p;
}

export function theta2(v: number): number {
  initTables();
  const c = Math.cos(2.0 * Math.PI * v);
  let p = 1.0;
  for (let j = 1; j <= NTERMS; j++) {
    const qh = _qhalf[j];
    p *= (1.0 - _qpow[j]) * (1.0 - 2.0 * c * qh + qh * qh);
  }
  return p;
}

export function theta3(v: number): number {
  initTables();
  const c = Math.cos(2.0 * Math.PI * v);
  let p = 1.0;
  for (let j = 1; j <= NTERMS; j++) {
    const qh = _qhalf[j];
    p *= (1.0 - _qpow[j]) * (1.0 + 2.0 * c * qh + qh * qh);
  }
  return p;
}

// ── E8 commitment (eq. 2.31) ─────────────────────────────────────────────

export interface E8Commit {
  pi1: number;   // ∏ θ₁(yₗ), l = 1..8
  pi2: number;   // ∏ θ₂(yₗ)
  pi3: number;   // ∏ θ₃(yₗ)
  C:   number;   // ½(pi1 + pi2 + pi3)
}

export function e8Commit(y: ArrayLike<number>): E8Commit {
  let pi1 = 1.0, pi2 = 1.0, pi3 = 1.0;
  for (let l = 0; l < E8_DIM; l++) {
    pi1 *= theta1(y[l]);
    pi2 *= theta2(y[l]);
    pi3 *= theta3(y[l]);
  }
  return { pi1, pi2, pi3, C: 0.5 * (pi1 + pi2 + pi3) };
}

// ── E8×E8 two-factor commitment (Section 4) ───────────────────────────────

export interface E8xE8Commit {
  c1: E8Commit;
  c2: E8Commit;
  C:  number;   // c1.C × c2.C
}

export function e8xe8Commit(
  y1: ArrayLike<number>,
  y2: ArrayLike<number>,
): E8xE8Commit {
  const c1 = e8Commit(y1);
  const c2 = e8Commit(y2);
  return { c1, c2, C: c1.C * c2.C };
}

// ── Demo hash: FNV-1a + xorshift64 (mirrors C exactly via BigInt) ─────────
// In production replace with passwordToRootsHKDF() below.

const FNV_OFFSET = 0xcbf29ce484222325n;
const FNV_PRIME  = 0x100000001b3n;
const MASK64     = 0xFFFFFFFFFFFFFFFFn;

function xorshift64(s: bigint): bigint {
  s ^= (s << 13n) & MASK64;
  s ^=  s >>  7n;            // right-shift: no overflow possible
  s ^= (s << 17n) & MASK64;
  return s;
}

/**
 * passwordToRoots — demo hash, identical algorithm to C's password_to_roots().
 * Produces bit-exact same roots as C for the same input bytes.
 *
 * The password is encoded to UTF-8 bytes first (matching C's char* iteration)
 * so that multi-byte Unicode characters hash identically in both languages.
 * DO NOT use in production — use passwordToRootsHKDF() instead.
 */
export function passwordToRoots(password: string): number[] {
  // Encode to UTF-8 bytes — matches C's byte-by-byte char* traversal
  const bytes = new TextEncoder().encode(password);
  let state = FNV_OFFSET;
  for (let i = 0; i < bytes.length; i++) {
    state ^= BigInt(bytes[i]);
    state  = (state * FNV_PRIME) & MASK64;
  }
  const roots: number[] = [];
  for (let l = 0; l < E8_DIM; l++) {
    state = xorshift64(state);
    // Mirror C: r = state >> 11, y = r / 2^53
    const r = Number(state >> 11n);
    let y = r / 9007199254740992; // 2^53
    if (y < 1e-9) y = 1e-9;
    roots.push(y);
  }
  return roots;
}

// ── Production HKDF path ─────────────────────────────────────────────────
// Uses WebCrypto (available in browser, Deno, and Node ≥ 19).

/**
 * Derive 8 Chern roots in (0,1) from a password and salt using HKDF-SHA256.
 * FIPS 198-compliant key derivation — use this in production.
 */
export async function passwordToRootsHKDF(
  password: string,
  salt: Uint8Array,
): Promise<number[]> {
  const enc = new TextEncoder();
  const ikm = enc.encode(password);
  const key = await crypto.subtle.importKey(
    'raw', ikm, 'HKDF', false, ['deriveBits'],
  );
  const raw = await crypto.subtle.deriveBits(
    {
      name: 'HKDF',
      hash: 'SHA-256',
      salt,
      info: enc.encode('e8-chern-roots-v1'),
    },
    key,
    E8_DIM * 64,   // 8 roots × 8 bytes
  );
  return rawBytesToRoots(new Uint8Array(raw));
}

/**
 * Derive 8 Chern roots from a record's integrity fields using HKDF-SHA256.
 * `fields` is the set of columns that must not change without detection.
 * `orgSalt` is an org-level secret (e.g. org_id + a server-side pepper).
 */
export async function recordToRoots(
  fields: Record<string, unknown>,
  orgSalt: string,
): Promise<number[]> {
  const enc = new TextEncoder();
  // Canonical serialisation: sorted keys → deterministic across platforms
  const canonical = JSON.stringify(fields, Object.keys(fields).sort());
  const key = await crypto.subtle.importKey(
    'raw', enc.encode(orgSalt), 'HKDF', false, ['deriveBits'],
  );
  const raw = await crypto.subtle.deriveBits(
    {
      name: 'HKDF',
      hash: 'SHA-256',
      salt: enc.encode(canonical),
      info: enc.encode('e8-record-integrity-v1'),
    },
    key,
    E8_DIM * 64,
  );
  return rawBytesToRoots(new Uint8Array(raw));
}

function rawBytesToRoots(arr: Uint8Array): number[] {
  const roots: number[] = [];
  for (let i = 0; i < E8_DIM; i++) {
    // Use top 53 bits of each 8-byte chunk → double in (0,1)
    const b = i * 8;
    const hi = (arr[b] << 24 | arr[b+1] << 16 | arr[b+2] << 8 | arr[b+3]) >>> 0;
    const lo = (arr[b+4] << 24 | arr[b+5] << 16 | arr[b+6] << 8 | arr[b+7]) >>> 0;
    const big = (BigInt(hi) << 32n) | BigInt(lo);
    const mantissa = Number(big >> 11n);          // 53 significant bits
    let y = mantissa / 9007199254740992;           // / 2^53
    if (y < 1e-9) y = 1e-9;
    if (y > 1.0 - 1e-9) y = 1.0 - 1e-9;
    roots.push(y);
  }
  return roots;
}

// ── Bit-level utilities for cross-language verification ───────────────────

/** Return the IEEE 754 bit pattern of a double as a 16-char hex string. */
export function doubleToHex(v: number): string {
  const buf = new ArrayBuffer(8);
  new DataView(buf).setFloat64(0, v, false); // big-endian
  return Array.from(new Uint8Array(buf), b => b.toString(16).padStart(2, '0')).join('');
}

/** Round-trip a hex string back to a double. */
export function hexToDouble(h: string): number {
  const arr = new Uint8Array(8);
  for (let i = 0; i < 8; i++) arr[i] = parseInt(h.slice(i*2, i*2+2), 16);
  return new DataView(arr.buffer).getFloat64(0, false);
}
